tests/update_state_test.py

```diff
@@ -0,0 +1,231 @@
+    '''Update_state sets guessing to True or False, then sets skip True first then False,
+    and finally, switches the team color and resets guess_value to zero'''
+    def test_guessing_is_true_under_right_conditions(self):
+        'If these three variables are as they are set here, guessing should be True'
+        game.guess_value = 1
+        game.code_number = 1
+        game.game_over = False
+
+        game.update_state()
+
+        assert game.guessing is True
+
+    def test_guessing_is_false_when_game_over_is_true(self):
+        'If game_over is True, guessing should be False'
+        game.guess_value = 1
+        game.code_number = 1
+        game.game_over = True
+        # need skip to be True so that self.guessing is not reset to True before test
+        game.skip = True
+
+        game.update_state()
+
+        assert game.guessing is False
+
+    def test_guessing_is_false_when_code_number_is_zero(self):
+        'If code_number is 0, guessing should be False'
+        game.guess_value = 1
+        game.code_number = 0
+        game.game_over = False
+        # need skip to be True so that self.guessing is not reset to True before test
+        game.skip = True
+
+        game.update_state()
+
+        assert game.guessing is False
+
+    def test_guessing_is_false_when_guess_value_is_greater_than_one(self):
+        'If guess_value is more than one, guessing should be False'
+        game.guess_value = 2
+        game.code_number = 1
+        game.game_over = False
+        # need skip to be True so that self.guessing is not reset to True before test
+        game.skip = True
+
+        game.update_state()
+
+        assert game.guessing is False
+
+    def test_skip_is_true_when_first_guessing_is_false(self):
+        'When guessing is false and skip is false, skip should be changed to True'
+        game.skip = False
+
+        game.update_state()
+
+        assert game.skip is True
+
+    def test_skip_is_false_when_second_guessing_is_false(self):
+        'When guessing is false and skip is True, skip should be changed to False'
+        game.skip = True
+
+        game.update_state()
+
+        assert game.skip is False
+
+    def test_skip_is_true_sets_guessing_to_true(self):
+        'When skip is True, guessing should be set equal to True'
+        # Set skip to false here as function will change it to True
+        game.skip = False
+
+        game.update_state()
+
+        assert game.guessing is True
+
+    def test_skip_is_false_so_team_red_goes_blue_and_guess_value_reset(self):
+        'When skip is False, the team should be switched and the guess_value reset'
+        # Set skip to True so that function will change it to False for the test
+        game.skip = True
+        game.team = 'red'
+        game.guess_value = 1
+
+        game.update_state()
+
+        assert game.team == 'blue'
+        assert game.guess_value == 0
+
+    def test_skip_is_false_so_team_blue_goes_red_and_guess_value_reset(self):
```

Comment:
This one seemed like a good example of what I mean by the previous value overriding what I try to set the new value as. Here, I set the value of `game.team` to `'blue'` update state will change it to `'red'`. 

When this test is run individually that is what happens and the test passes. 

When I run all the tests together, this test fails as `game.team` is `'blue'` not `'red'`. This could only happen if the original value of `game.team` is `'red'` and so `update_state()` changes it to `'blue'`.  Which would mean that my setting it to `'blue'` at the start of this new test does nothing when I run the tests together. Do you know why this may be happening?

Reply:
This is because you have `game = Game()` at the global scope (outside your functions). If you run the tests individually, `game` has just been initialized, but if you run the tests all together, the state of `game` has been affected by all the tests that come before `test_skip_is_false_so_team_blue_goes_red_and_guess_value_reset`.

To see this in action, temporarily put `print(f"code_number={game.code_number}")` right before your call to game.update_state()`. Then call `pytest -s` from the `tests` directory. Then add `game = Game()` to the start of the test function and rerun the tests. This test function is using the initialization value of `code_number`, which is `0`, if you run the test individually, but if you run the tests in a group, `test_guessing_is_false_when_guess_value_is_greater_than_one` sets `code_number = 1`.

![image](https://github.com/LKieser/codenames/assets/5274060/6e45982b-2abe-4a65-8feb-5d18176e4698)
![image](https://github.com/LKieser/codenames/assets/5274060/28ba142d-c9fd-48e2-9fa7-41124e18e202)

Your unit tests should be independent of each other, AKA the changes one tests makes should not affect the next test. Each test function should call `game = Game()` to start the test in a known state and you should be able to run all your tests together and get the same results as when you run them individually.